runtime/pallene_core.c

```diff
@@ -23,64 +24,47 @@ const char *pallene_tag_name(int raw_tag)
     }
 }
 
-void pallene_runtime_arity_error(
-    lua_State *L,
-    int expected,
-    int received
-){
-    luaL_error(
-        L,
-        "wrong number of arguments to function, expected %d but received %d",
-        expected, received
-    );
-    PALLENE_UNREACHABLE;
-}
-
-void pallene_runtime_argument_type_error(
+void pallene_runtime_tag_check_error(
     lua_State *L,
-    const char *param_name,
     int line,
     int expected_tag,
-    TValue *slot
-){
-    const char *expected_type = pallene_tag_name(expected_tag);
-    const char *received_type = pallene_tag_name(rawtt(slot));
-    luaL_error(
-        L,
-        "wrong type for argument %s at line %d, expected %s but found %s",
-        param_name, line, expected_type, received_type
-    );
-    PALLENE_UNREACHABLE;
-}
-
-void pallene_runtime_array_type_error(
-   lua_State *L,
-   int line,
-   int expected_tag,
-   int received_tag
+    int received_tag,
+    const char *description_fmt,
+    ...
 ){
     const char *expected_type = pallene_tag_name(expected_tag);
     const char *received_type = pallene_tag_name(received_tag);
-    luaL_error(
-        L,
-        "wrong type for array element at line %d, expected %s but found %s",
-        line, expected_type, received_type
-    );
+
+    // This implementation is inspired by luaL_error
+    luaL_where(L, 1);
+    if (line > 0) {
+        lua_pushfstring(L, "line %d: ", line);
+    } else {
+        lua_pushfstring(L, "");
+    }
+    lua_pushfstring(L, "wrong type for ");
+    {
+        va_list argp;
+        va_start(argp, description_fmt);
+        lua_pushvfstring(L, description_fmt, argp);
+        va_end(argp);
+    }
+    lua_pushfstring(L, ", expected %s but found %s",
+        expected_type, received_type);
+    lua_concat(L, 5);
+    lua_error(L);
     PALLENE_UNREACHABLE;
 }
 
-void pallene_runtime_function_return_error(
+void pallene_runtime_arity_error(
     lua_State *L,
-    int line,
-    int expected_tag,
-    int received_tag
+    int expected,
+    int received
 ){
-    const char *expected_type = pallene_tag_name(expected_tag);
-    const char *received_type = pallene_tag_name(received_tag);
     luaL_error(
         L,
-        "wrong type for function result at line %d, expected %s but found %s",
-        line, expected_type, received_type
+        "wrong number of arguments to function, expected %d but received %d",
```

Comment:
We could add the function name and the line it is declared.

Reply:
This belongs in a separate PR

Reply:
The commit diff is a bit misleading. The original arity_error function was not modified by this PR, it only moved to a different place in the file.